config/dynamic.c

```diff
@@ -16,11 +16,467 @@
+		return -1;
+	}
+
+	/* Function to be called. */
+	lua_getglobal(dy_conf->lua_state, "dy_cmd_proc");
+	lua_pushlightuserdata(dy_conf->lua_state, reply_msg);
+	ret = lua_pcall(dy_conf->lua_state, 1, 1, 0);
+	if (ret != 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: %s!\n", lua_tostring(dy_conf->lua_state, -1));
+		return -1;
+	}
+
+	ret = luaL_checkinteger(dy_conf->lua_state, -1);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"dyn_cfg: dy_cmd_proc() return value is %d!\n",
+			ret);
+		reply_client_message(conn_fd,
+			CLIENT_PROC_ERROR, strlen(CLIENT_PROC_ERROR));
+		return -1;
+	}
+
+	reply_len = strlen(reply_msg);
+	if (reply_len > LUA_MSG_MAX_LEN) {
+		RTE_LOG(ERR, MALLOC,
+			"dyn_cfg: The reply message length (%d) exceeds the limit!\n",
+			reply_len);
+		reply_client_message(conn_fd,
+			CLIENT_PROC_ERROR, strlen(CLIENT_PROC_ERROR));
+		return -1;
+	}
+
+	return reply_client_message(conn_fd, reply_msg, reply_len);
+}
+
+/*
+ * Return values:
+ * 	-1: Error happens.
+ * 	0: Command successfully processed, may need to process further commands.
+ * 	1: Connection closed by the client.
+ */
+static int
+process_single_cmd(int conn_fd, char *msg_buff,
+	const char *client_ip, struct dynamic_config *dy_conf)
+{
+	int recv_size;
+	int tot_size = 0;
+	uint16_t msg_len = 0;
+
+	/*
+	 * The protocol should be rather simple: two-byte,
+	 * unsigned integer in network order signal the size,
+	 * in bytes, of the message that follows that
+	 * first two bytes.
+	 */
+	while ((recv_size = recv(conn_fd,
+			(char *)&msg_len + tot_size, 1, 0)) > 0) {
+		tot_size += recv_size;
+		if (tot_size >= 2)
+			break;
+	}
+
+	/* The connection with the client is closed. */
+	if (recv_size == 0) {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"dyn_cfg: Client %s disconnected!\n",
+			client_ip);
+		return 1;
+	}
+
+	msg_len = ntohs(msg_len);
+	if (msg_len > LUA_MSG_MAX_LEN) {
+		RTE_LOG(ERR, MALLOC,
+			"dyn_cfg: The message length (%d) exceeds the limit!\n",
+			msg_len);
+		return -1;
+	}
+
+	tot_size = 0;
+	memset(msg_buff, 0, sizeof(char) * (LUA_MSG_MAX_LEN + 1));
+	while ((recv_size = recv(conn_fd, msg_buff + tot_size,
+			msg_len - tot_size, 0)) > 0) {
+		tot_size += recv_size;
+		if (tot_size >= msg_len) {
+			/*
+			 * Error in handling client message,
+			 * close the connection.
+			 */
+			int ret = process_client_message(
+				conn_fd, msg_buff, msg_len, dy_conf);
+			if (ret < 0)
+				return -1;
+
+			break;
+		}
+	}
+
+	/* The connection with the client is closed. */
+	if (recv_size == 0) {
+		RTE_LOG(NOTICE, GATEKEEPER,
+			"dyn_cfg: Client %s disconnected!\n",
+			client_ip);
+		return 1;
+	}
+
+	return 0;
+}
+
+static void
+handle_client(int server_socket_fd, struct dynamic_config *dy_conf)
+{
+	int ret;
+	int conn_fd;
+	socklen_t len;
+	struct sockaddr_storage client_addr;
+	char client_ip[INET6_ADDRSTRLEN];
+	int client_port;
+	char *msg_buff = rte_calloc(
+			"msg_buff", LUA_MSG_MAX_LEN + 1, sizeof(char), 0);
```

Comment:
Why don't you allocate `msg_buff` in the stack, i.e. as a local variable? Besides producing simpler code, it would solve memory leaks that happen below when an error happens and the function returns without proper cleanup.